src/awkward/_v2/contents/indexedoptionarray.py

```diff
@@ -578,6 +578,210 @@ def _localindex(self, axis, depth):
+
+        next_length = index_length - numnull[0]
+        nextparents = ak._v2.index.Index64.empty(next_length, nplike)
+        nextcarry = ak._v2.index.Index64.empty(next_length, nplike)
+        outindex = ak._v2.index.Index64.empty(index_length, nplike)
+        self._handle_error(
+            nplike[
+                "awkward_IndexedArray_reduce_next_64",
+                nextcarry.dtype.type,
+                nextparents.dtype.type,
+                outindex.dtype.type,
+                self._index.dtype.type,
+                parents.dtype.type,
+            ](
+                nextcarry.to(nplike),
+                nextparents.to(nplike),
+                outindex.to(nplike),
+                self._index.to(nplike),
+                parents.to(nplike),
+                index_length,
+            )
+        )
+        next = self._content._carry(nextcarry, False, NestedIndexError)
+        out = next._unique(
+            negaxis,
+            starts,
+            nextparents,
+            outlength,
+        )
+
+        if branch or (negaxis is not None and negaxis != depth):
+            nextoutindex = ak._v2.index.Index64.empty(parents_length, nplike)
+            self._handle_error(
+                nplike[
+                    "awkward_IndexedArray_local_preparenext_64",
+                    nextoutindex.dtype.type,
+                    starts.dtype.type,
+                    parents.dtype.type,
+                    nextparents.dtype.type,
+                ](
+                    nextoutindex.to(nplike),
+                    starts.to(nplike),
+                    parents.to(nplike),
+                    parents_length,
+                    nextparents.to(nplike),
+                    next_length,
+                )
+            )
+
+            out = ak._v2.contents.IndexedOptionArray(
+                nextoutindex,
+                out,
+                None,
+                self._parameters,
+            ).simplify_optiontype()
+
+            return out
+
+        if isinstance(out, ak._v2.contents.ListOffsetArray):
+            # FIXME: move to kernel
+            nulls = []
+            for i in range(len(self._index)):
+                if self._index[i] < 0:
+                    nulls.append(parents[i])
+
+            newindex = ak._v2.index.Index64.zeros(out._offsets[-1] + len(nulls), nplike)
+            newoffsets = ak._v2.index.Index64.zeros(len(out._offsets), nplike)
+            k = 0
+            ll = 0
+            shift = 0
+            newindex[0] = ll
+            newoffsets[0] = out._offsets[0]
+            for i in range(len(starts)):
+                for _j in range(out._offsets[i], out._offsets[i + 1]):
+                    newindex[k] = ll
+                    k += 1
+                    ll += 1
+                if parents[starts[i]] in nulls:
+                    newindex[k] = -1
+                    k += 1
+                    shift += 1
+                newoffsets[i + 1] = out._offsets[i + 1] + shift
```

Comment:
Reminder to move to a kernel (can resolve this comment when done).

Reply:
done